gtda/externals/bindings/ripser_bindings.cpp

```diff
@@ -32,21 +34,22 @@ PYBIND11_MODULE(gtda_ripser, m) {
       .def_readwrite("num_edges", &ripserResults::num_edges);
 
   m.def("rips_dm",
-        [](std::vector<float> D, int N, int modulus, int dim_max,
+        [](py::array_t<float>& D, int N, int modulus, int dim_max,
            float threshold, int do_cocycles) {
-          ripserResults ret =
-              rips_dm(&D[0], N, modulus, dim_max, threshold, do_cocycles);
+          ripserResults ret = rips_dm((float*)D.request().ptr, N, modulus,
```

Comment:
I'm sure it's right but it's too much for my brain, what does this line do?

Reply:
It's related on `pybind11` underlying layer to manage objects. `numpy` in `pybind11` is a special type, using `pybind11` classes to represent it. With `request().ptr` I ask to have access to the underlying data inside the `object`.

It's not easy to see this in the `pybind11` documentation, the API is currently not well documented in my opinion.

Reply:
I see. And is `(float*)` about passing a pointer instead of an object?

Reply:
Well that's something we unfortunately inherit from `ripser.py`, but yeah you're right

Reply:
I thought that this was the core of your intervention in this PR! Could you explain where exactly you change the behaviour so that pointers are passed and memory is saved?

Reply:
Also what's the difference between the previous `&` syntax and the new `(float*)` one?

Reply:
The `&` syntax is passing by reference, it's a c++ syntax, one of it's useful uses is to prevent the copy of an object.
The `float*` syntax is to have a pointer to the memory of the data, it's more a C syntax and in C++ we try to move away from it.

Reply:
For the first question:

I prevent copying the memory from Python to C++ with the `&` syntax (It's a bit more complicated than that, but to have the big picture).
I then need to have an access to the raw pointer of the data `float*` because it's a requirement from the C++ ripser in `ripser.py`

Reply:
So I guess my question is why the `&` was not sufficient, if it already prevented copies which was the purpose of this PR?